### PR TITLE
Fixed #32813 -- Made runserver display port after binding.

### PR DESCRIPTION
**Pull request by Dhanush-arch**
**Fixed #32813** - Display development server address after server bind

If a user selects port ```0``` for runserver command, which is a wildcard port:

```
$ ./manage.py runserver 127.0.0.1:0
...
Starting development server at http://127.0.0.1:0/
Quit the server with CONTROL-C.
```
The app will not be running in the given link ```http://127.0.0.1:0/```, because the operating system selects a port.

After this change Django will be able to display the correct port:

```
$ ./manage.py runserver 127.0.0.1:0
...
Starting development server at http://127.0.0.1:37337/
Quit the server with CONTROL-C.
```
The original PR [#14250](https://github.com/django/django/pull/14250)  was closed. The current PR [#15748](https://github.com/django/django/pull/15748) for the ticket has been inactive and also requires some changes due to the merged PR [#16145](https://github.com/django/django/pull/16145). I have added the changes and updated the test case.

### CODE REVIEW

File: django/core/management/commands/runserver.py [hunk 1]
```diff
@@ -168,23 +143,52 @@ def inner_run(self, *args, **options):
...
-            self.stderr.write("Error: %s" % error_text)
+            self.stderr.write(f"Error: {error_text}")
             # Need to use an OS exit because sys.exit doesn't work in a thread
             os._exit(1)
         except KeyboardInterrupt:
             if shutdown_message:
                 self.stdout.write(shutdown_message)
             sys.exit(0)
+
+    def on_bind(self, server_port):
```
Comment (apollo13) [inline django/core/management/commands/runserver.py:168]: I wonder if it would make sense to also pass in `server_name` here which is also set by `http.server` and use that instead of `self.addr` -- might get rid of the ifs below.

**Review by apollo13 — COMMENTED** [review]
My comments are mostly nitpicks, I don't think that the PR is wrong per se but could be improved stylewise. I'll let @felixxm be the judge :)

File: django/core/servers/basehttp.py [hunk 2]
```diff
@@ -227,13 +227,21 @@ def handle_one_request(self):
...
     server_address = (addr, port)
-    if threading:
-        httpd_cls = type("WSGIServer", (socketserver.ThreadingMixIn, server_cls), {})
-    else:
-        httpd_cls = server_cls
+    bases = (socketserver.ThreadingMixIn, server_cls) if threading else (server_cls,)
+    httpd_cls = type("WSGIServer", bases, {})
     httpd = httpd_cls(server_address, WSGIRequestHandler, ipv6=ipv6)
+    if on_bind is not None:
+        on_bind(httpd.server_port)
```
Comment (apollo13) [inline django/core/servers/basehttp.py:244]: Btw if we are afraid that some `server_cls`  implementations might not expose `server_port/server_name` we could hook into the base environ where this needs to be set I think (though it is questionable if that is exposed under the same name)

File: django/core/servers/basehttp.py [hunk 3]
```diff
@@ -227,13 +227,21 @@ def handle_one_request(self):
...
+    on_bind=None,
+    server_cls=WSGIServer,
+):
     server_address = (addr, port)
-    if threading:
-        httpd_cls = type("WSGIServer", (socketserver.ThreadingMixIn, server_cls), {})
-    else:
-        httpd_cls = server_cls
+    bases = (socketserver.ThreadingMixIn, server_cls) if threading else (server_cls,)
+    httpd_cls = type("WSGIServer", bases, {})
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/core/servers/basehttp.py:241]: Moreover, it's unrelated refactoring, please revert.

File: django/core/management/commands/runserver.py [hunk 4]
```diff
@@ -168,23 +143,52 @@ def inner_run(self, *args, **options):
...
+                errno.EACCES: f"You don't have permission to access port {self.port}.",
+                errno.EADDRINUSE: f"Port {self.port} is already in use.",
+                errno.EADDRNOTAVAIL: f"IP address {self.addr!r} can't be assigned to.",
             }
             try:
                 error_text = ERRORS[e.errno]
             except KeyError:
                 error_text = e
-            self.stderr.write("Error: %s" % error_text)
+            self.stderr.write(f"Error: {error_text}")
```
Comment (felixxm) [inline django/core/management/commands/runserver.py:160]: Please revert.

File: django/core/management/commands/runserver.py [hunk 5]
```diff
@@ -168,23 +143,52 @@ def inner_run(self, *args, **options):
...
             )
         except OSError as e:
             # Use helpful error messages instead of ugly tracebacks.
             ERRORS = {
-                errno.EACCES: "You don't have permission to access that port.",
-                errno.EADDRINUSE: "That port is already in use.",
-                errno.EADDRNOTAVAIL: "That IP address can't be assigned to.",
+                errno.EACCES: f"You don't have permission to access port {self.port}.",
+                errno.EADDRINUSE: f"Port {self.port} is already in use.",
+                errno.EADDRNOTAVAIL: f"IP address {self.addr!r} can't be assigned to.",
```
Comment (felixxm) [inline django/core/management/commands/runserver.py:154]: This looks like an unrelated cleanup, that should be done in a separate PR. It's also untested.

File: django/core/servers/basehttp.py (hunk 2 above)
Comment (felixxm) [reply django/core/servers/basehttp.py:246]: Can we use `port` passed to `run()` as a fallback?
```suggestion
        on_bind(getattr(httpd, "server_port", port))
```

File: django/core/management/commands/runserver.py (hunk 1 above)
Comment (Dhanush-arch) [reply django/core/management/commands/runserver.py:168]: Can you explain how `httpd.server_name` can be used instead of `self.addr` ? I guess I did not understand this part. Because `httpd.server_name` has its value as `localhost` or the host name of my system, depending on the IP addr provided.

File: django/core/management/commands/runserver.py [hunk 6]
```diff
@@ -168,23 +143,52 @@ def inner_run(self, *args, **options):
...
+
+        if self._raw_ipv6:
+            addr = f"[{self.addr}]"
+        elif self.addr == "0":
+            addr = "0.0.0.0"
+        else:
+            addr = self.addr
+
+        now = datetime.now().strftime("%B %d, %Y - %X")
+        self.stdout.write(now)
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/core/management/commands/runserver.py]: I'd leave it for a separate cleanup.

**Comment by felixxm** [comment]
@apollo13 Can you take another look? :star2:

File: django/core/management/commands/runserver.py (hunk 1 above)
Comment (apollo13) [reply django/core/management/commands/runserver.py:168]: Ah okay, I thought it would be the resolved/supplied address

**Review by apollo13 — APPROVED** [review]

**Comment by felixxm** [comment]
@Dhanush-arch Thanks :+1

@apollo13 Thanks for reviews :+1:
